### PR TITLE
Fix llvm-assertion: don't instantiate 0-sized structs (ghost types)

### PR DESCRIPTION
**Pull request by NHDaly**
I created this commit from a patch that @JeffBezanson sent me. I'm not familiar with this code myself, so I don't know if this is sufficient, but I wanted to open this PR so that we didn't forget about this fix! :) Feel free to commit to it directly / take it over / replace it with a new PR or whatever.

In particular, we were a bit unsure about what should go in the second `else` block, where we currently put `return e;`.

------------------

As I understand it, LLVM doesn't support 0-sized structs (ghost types),
which julia makes significant use of. This fixes an incorrect behavior
in the julia compiler accidentally attempting to instantiate such
0-sized structs.

When using a debug build of LLVM, this fails with an assertion error,
but failed silently on standard LLVM.

-----------------------------------

This came up when debugging https://github.com/JuliaLang/julia/issues/31418.

### CODE REVIEW

**Comment by NHDaly** [comment]
Also i think this should get some tests as well.

File: src/intrinsics.cpp [hunk 1]
```diff
@@ -600,8 +600,13 @@ static jl_cgval_t emit_pointerref(jl_codectx_t &ctx, jl_cgval_t *argv)
...
         Type *ptrty = julia_type_to_llvm(ety, &isboxed);
         assert(!isboxed);
-        Value *thePtr = emit_unbox(ctx, ptrty->getPointerTo(), e, e.typ);
-        return typed_load(ctx, thePtr, im1, ety, tbaa_data, nullptr, true, align_nb);
+        if (!type_is_ghost(ptrty)) {
+            Value *thePtr = emit_unbox(ctx, ptrty->getPointerTo(), e, e.typ);
+            return typed_load(ctx, thePtr, im1, ety, tbaa_data, nullptr, true, align_nb);
+        }
+        else {
+            return mark_julia_const(((jl_datatype_t*)ety)->instance);
```
Comment (vtjnash) [inline src/intrinsics.cpp]:
```suggestion
            return ghostValue(ety);
```

File: src/intrinsics.cpp [hunk 2]
```diff
@@ -663,8 +668,12 @@ static jl_cgval_t emit_pointerset(jl_codectx_t &ctx, jl_cgval_t *argv)
         bool isboxed;
         Type *ptrty = julia_type_to_llvm(ety, &isboxed);
         assert(!isboxed);
-        thePtr = emit_unbox(ctx, ptrty->getPointerTo(), e, e.typ);
-        typed_store(ctx, thePtr, im1, x, ety, tbaa_data, nullptr, nullptr, align_nb);
+        if (!type_is_ghost(ptrty)) {
+            thePtr = emit_unbox(ctx, ptrty->getPointerTo(), e, e.typ);
+            typed_store(ctx, thePtr, im1, x, ety, tbaa_data, nullptr, nullptr, align_nb);
+        } else {
+          return e;
```
Comment (vtjnash) [inline src/intrinsics.cpp]: I suggest just letting this fall-through (no else clause), then replace the `mark_julia_type` at the end with simply `return e`

**Comment by vtjnash** [comment]
fixes #29929

File: src/intrinsics.cpp (hunk 2 above)
Comment (NHDaly) [reply src/intrinsics.cpp]: Done.

**Comment by NHDaly** [comment]
I applied your suggestions. I don't know what this code does so i'm acting blind. :)

**Comment by NHDaly** [comment]
I've added a test case for #29929 and a few others that exercise this change.

**Review by JeffBezanson — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
Any reason not to merge? This PR appears to have a higher percentage of green checkmarks that I no longer know the meaning of than the average PR.

**Comment by JeffBezanson** [comment]
Should have squashed!!

**Comment by vchuravy** [comment]
Mea culpa!

**Pull request merged**